sources/web/datalab/polymer/test/table-inline-details-test.ts

```diff
@@ -0,0 +1,159 @@
+
+      const response = {
+        result: mockTable,
+      };
+      return Promise.resolve(response) as
+          Promise<gapi.client.HttpRequestFulfilled<gapi.client.bigquery.Table>>;
+    };
+    GapiManager.bigquery.getTableRows = (pid, did, tid, _maxResults) => {
+      assert(!!pid && !!did && !!tid,
+          'getTableDetails should be called with project, dataset, and table');
+
+      const response = {
+        result: mockTabledata,
+      } as gapi.client.HttpRequestFulfilled<ListTabledataResponse>;
+      return Promise.resolve(response) as
+          Promise<gapi.client.HttpRequestFulfilled<gapi.client.bigquery.ListTabledataResponse>>;
+    };
+
+    testFixture = fixture('table-inline-details-fixture');
+    testFixture.tableId = '';
+    Polymer.dom.flush();
+  });
+
+  it('displays an empty element if no table is provided', () => {
+    assert(!testFixture.$.placeholder.hidden, 'placeholder should be shown');
+    assert(testFixture.$.container.hidden, 'container should be hidden');
+    assert(testFixture.$.placeholder.querySelector('paper-spinner'),
+        'spinner should be hidden if no table is loading');
+  });
+
+  it('loads the table schema', (done: () => null) => {
+    testFixture.addEventListener('_table-changed', () => {
+      Polymer.dom.flush();
+
+      assert(JSON.stringify((testFixture as any)._table) === JSON.stringify(mockTable),
+          'element should have fetched table info');
+
+      const schemaRow = testFixture.$.tabledata.rows[0];
+      const columns = schemaRow.querySelectorAll('th');
+      assert(columns.length === mockTable.schema.fields.length,
+          'unexpected number of schema rows');
+
+      mockTable.schema.fields.forEach((field, i) => {
+        const spans = schemaRow.children[i].querySelectorAll('span');
+        assert(spans[0].innerText === field.name, 'field name does not match');
+        assert(spans[1].innerText === field.type, 'field type does not match');
+      });
+
+      done();
+    });
+
+    testFixture.tableId = 'pid:did.tid';
+  });
+
+  it('loads the table data', (done: () => null) => {
+    testFixture.addEventListener('_rows-changed', () => {
+      Polymer.dom.flush();
+
+      assert(JSON.stringify((testFixture as any)._rows) ===
+             JSON.stringify(mockTabledata.rows),
+          'element should have fetched table rows');
+
+      const dataRows = testFixture.$.tabledata.querySelectorAll('tr.data');
+      assert(dataRows.length === mockTabledata.rows.length,
+          'unexpected number of schema rows');
+
+      mockTabledata.rows.forEach((row, r) => {
+        const dataFields = dataRows[r].querySelectorAll('td');
+        assert(dataFields.length === row.f.length,
+            'column count does not match');
+        row.f.forEach((field, f) => {
+          assert(field.v === dataFields[f].innerText, 'field data does not match');
+        });
+      });
+
+      done();
+    });
+
+    testFixture.tableId = 'pid:did.tid';
+  });
+
```

Comment:
I'd add another test for when data fails to load.

Reply:
The code doesn't really do anything when the data fails to load, so I'm not sure what I would be testing for. I added a TODO to add some tests once the code handles errors better.